File: app/routes/api_v2/schemas/survey.py
"""Anket sistemi Pydantic şemaları (Faz 1)."""
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


# --- Katalog ---

class SurveyTemplateBrief(BaseModel):
    id: int
    code: str
    title: str
    description: str = ""
    category: str
    category_label: str
    scoring_type: str
    question_count: int = 0
    estimated_minutes: int = 10
    source_attribution: str = ""


class SurveyCatalogResponse(BaseModel):
    items: list[SurveyTemplateBrief]
    categories: dict[str, str]  # key → TR etiket


# --- Sorular / doldurma ---

class SurveyQuestionModel(BaseModel):
    id: int
    order_no: int
    text: str
    qtype: str
    dimension_key: str | None = None
    options: list[dict] = Field(default_factory=list)


# --- Sonuç ---

class SurveyDimensionScore(BaseModel):
    key: str
    label: str
    description: str = ""
    score_pct: float
    level: str
    level_label: str
    high_is_good: bool = True
    comment: str = ""


class SurveyQualitativeBlock(BaseModel):
    key: str
    label: str
    description: str = ""
    entries: list[dict] = Field(default_factory=list)


class SurveyResultModel(BaseModel):
    scoring_type: str
    dimensions: list[SurveyDimensionScore] = Field(default_factory=list)
    top_dimensions: list[str] = Field(default_factory=list)
    qualitative: list[SurveyQualitativeBlock] = Field(default_factory=list)
    open_answers: list[dict] = Field(default_factory=list)
    report_note: str = ""
    source_attribution: str = ""
    disclaimer: str


# --- Atamalar ---

class SurveyAssignmentRow(BaseModel):
    id: int
    template: SurveyTemplateBrief
    status: str
    status_label: str
    note: str = ""
    assigned_at: datetime
    started_at: datetime | None = None
    completed_at: datetime | None = None
    teacher_name: str | None = None
    student_name: str | None = None
    answered_count: int = 0


class SurveyAssignmentDetail(BaseModel):
    assignment: SurveyAssignmentRow
    result: SurveyResultModel | None = None


# --- Koç yüzeyi ---

class TeacherStudentSurveysResponse(BaseModel):
    assignments: list[SurveyAssignmentRow]
    catalog: list[SurveyTemplateBrief]
    categories: dict[str, str]


class SurveyAssignBody(BaseModel):
    template_id: int
    note: str = Field(default="", max_length=500)


class SurveyAssignResult(BaseModel):
    ok: bool = True
    assignment_id: int


class SurveyCancelResult(BaseModel):
    ok: bool = True


# --- AI Kariyer Sentezi ---

class CareerSuggestionModel(BaseModel):
    title: str
    field: str = ""
    why: str = ""
    example_departments: list[str] = Field(default_factory=list)


class CareerSynthesisModel(BaseModel):
    summary: str
    career_suggestions: list[CareerSuggestionModel] = Field(default_factory=list)
    strengths: list[str] = Field(default_factory=list)
    agenda: list[str] = Field(default_factory=list)
    watch_outs: list[str] = Field(default_factory=list)
    based_on_surveys: list[str] = Field(default_factory=list)  # anket başlıkları
    exam_count: int = 0
    generated_at: datetime | None = None


class CareerSynthesisCacheResponse(BaseModel):
    insight: CareerSynthesisModel | None = None
    is_stale: bool = False
    # Üretim için zorunlu anketler tamam mı; eksikse hangileri (başlıklar)
    ready: bool = False
    missing_surveys: list[str] = Field(default_factory=list)
    disclaimer: str


# --- Öğrenci yüzeyi ---

class StudentSurveysResponse(BaseModel):
    pending: list[SurveyAssignmentRow]
    completed: list[SurveyAssignmentRow]


class StudentSurveyFillResponse(BaseModel):
    assignment: SurveyAssignmentRow
    questions: list[SurveyQuestionModel]
    answers: dict[str, Any] = Field(default_factory=dict)
    result: SurveyResultModel | None = None  # tamamlandıysa öğrenci de görür
    disclaimer: str


class StudentSurveyAnswersBody(BaseModel):
    answers: dict[str, Any] = Field(default_factory=dict)
    complete: bool = False


class StudentSurveySaveResult(BaseModel):
    ok: bool = True
    status: str
    completed: bool = False
    missing_question_ids: list[int] = Field(default_factory=list)

File: web/lib/types/survey.ts
/** Anket sistemi tipleri — backend schemas/survey.py birebir karşılığı. */

export interface SurveyTemplateBrief {
  id: number;
  code: string;
  title: string;
  description: string;
  category: string;
  category_label: string;
  scoring_type: "dimensions" | "wheel" | "qualitative";
  question_count: number;
  estimated_minutes: number;
  source_attribution: string;
}

export interface SurveyCatalogResponse {
  items: SurveyTemplateBrief[];
  categories: Record<string, string>;
}

export interface SurveyQuestionModel {
  id: number;
  order_no: number;
  text: string;
  qtype: "likert5" | "slider10" | "choice" | "open";
  dimension_key: string | null;
  options: Array<{ value: string; label: string }>;
}

export interface SurveyDimensionScore {
  key: string;
  label: string;
  description: string;
  score_pct: number;
  level: "high" | "mid" | "low";
  level_label: string;
  high_is_good: boolean;
  comment: string;
}

export interface SurveyQualitativeBlock {
  key: string;
  label: string;
  description: string;
  entries: Array<{ question: string; answer: string }>;
}

export interface SurveyResultModel {
  scoring_type: "dimensions" | "wheel" | "qualitative";
  dimensions: SurveyDimensionScore[];
  top_dimensions: string[];
  qualitative: SurveyQualitativeBlock[];
  open_answers: Array<{ question: string; answer: string }>;
  report_note: string;
  source_attribution: string;
  disclaimer: string;
}

export interface SurveyAssignmentRow {
  id: number;
  template: SurveyTemplateBrief;
  status: "pending" | "in_progress" | "completed" | "cancelled";
  status_label: string;
  note: string;
  assigned_at: string;
  started_at: string | null;
  completed_at: string | null;
  teacher_name: string | null;
  student_name: string | null;
  answered_count: number;
}

export interface SurveyAssignmentDetail {
  assignment: SurveyAssignmentRow;
  result: SurveyResultModel | null;
}

export interface TeacherStudentSurveysResponse {
  assignments: SurveyAssignmentRow[];
  catalog: SurveyTemplateBrief[];
  categories: Record<string, string>;
}

export interface SurveyAssignBody {
  template_id: number;
  note?: string;
}

export interface SurveyAssignResult {
  ok: boolean;
  assignment_id: number;
}

export interface StudentSurveysResponse {
  pending: SurveyAssignmentRow[];
  completed: SurveyAssignmentRow[];
}

export interface StudentSurveyFillResponse {
  assignment: SurveyAssignmentRow;
  questions: SurveyQuestionModel[];
  answers: Record<string, number | string>;
  result: SurveyResultModel | null;
  disclaimer: string;
}

export interface CareerSuggestion {
  title: string;
  field: string;
  why: string;
  example_departments: string[];
}

export interface CareerSynthesisModel {
  summary: string;
  career_suggestions: CareerSuggestion[];
  strengths: string[];
  agenda: string[];
  watch_outs: string[];
  based_on_surveys: string[];
  exam_count: number;
  generated_at: string | null;
}

export interface CareerSynthesisCacheResponse {
  insight: CareerSynthesisModel | null;
  is_stale: boolean;
  ready: boolean;
  missing_surveys: string[];
  disclaimer: string;
}

export interface StudentSurveySaveResult {
  ok: boolean;
  status: string;
  completed: boolean;
  missing_question_ids: number[];
}
